### ISSUE TITLE
Weights in Custom Layer not trainable

### ISSUE THREAD

**Issue by braemd** [rstudio/keras#1123]
Hi,

I try to create a custom layer including other keras layers. I wanted to try it with just a dense layer as in this issue #926 but with no other layers in the model but the custom layer. I trained it on quite simple dummy data but in my example the weights not only were not present in the model summary but they also were not trained.

```
CustomDense <- R6::R6Class(
  "CustomDense",
  inherit = KerasLayer,
  lock_objects = FALSE,
  
   public = list(
    initialize = function(units) {
      self$units <- units
    },
    
    build = function(input_shape) {
      self$dense <- keras::layer_dense(units = self$units)
      super$build(input_shape)
    },
    
    call = function(input, mask = NULL) {
      output <- self$dense(input) 
    },
    
    compute_output_shape = function(input_shape) {
      input_shape[[2]] <- self$units 
      input_shape
    }
  )
)

custom_layer_dense <- function(object, 
                               units,
                               name = NULL, 
                               trainable = TRUE) {
  create_layer(CustomDense, 
               object, 
               args = list(units = units,
                           name = name,
                           trainable = trainable))
}

library(keras)
in_layer <- layer_input(shape=c(10L))
mid_layer <- custom_layer_dense(units=1L)
out_layer <- in_layer %>%
  mid_layer

model_custom <- keras_model(inputs=in_layer,outputs=out_layer)

model_custom %>% compile(
  optimizer = 'adam',
  loss = 'mae'
)

x_train <- array(rep(1,100),dim=c(10,10))
y_train <- array(rep(2,10),dim=c(10,1))

model_custom %>% fit(x_train,
                     y_train,
                     epochs=20)

```

During the training, the loss value stays exactly the same and the weights of the layer_dense do not change. But you can call the model_custom and you receive output.

```
>predict_on_batch(model_custom,x_train)
[,1]  [1,] -0.2629238  [2,] -0.2629238  [3,] -0.2629238  [4,] -0.2629238  [5,] -0.2629238  [6,] -0.2629238  [7,] -0.2629238  [8,] -0.2629238  [9,] -0.2629238 [10,] -0.2629238

```

Do I miss something important with custom layers? Or what might this problem solve?

Thanks in advance!

Note: Also the workaround with reticulate::PyClass did not work

Session Info:
R version 3.6.1
keras_2.3.0.0
tensorflow_2.2.0
reticulate_1.15

**Comment by dfalbel** [comment]
Actually, this is not expected to work because the weights won't be added to the layer's `trainable_weights`.
If using layers inside, it's better to write a custom model:

``` r
library(keras)

custom_layer_dense <- function(units, name = NULL) {
  keras_model_custom(name, model_fn = function(self) {
    self$dense <- keras::layer_dense(units = units)
    function(input, mask = NULL, training = FALSE) {
      self$dense(input)
    }
  })  
}

in_layer <- layer_input(shape=c(10L))
mid_layer <- custom_layer_dense(units=1L)
out_layer <- in_layer %>%
  mid_layer

model_custom <- keras_model(inputs=in_layer,outputs=out_layer)

model_custom %>% compile(
  optimizer = 'adam',
  loss = 'mae'
)

summary(model_custom)
#> Model: "functional_1"
#> ________________________________________________________________________________
#> Layer (type)                        Output Shape                    Param #     
#> ================================================================================
#> input_1 (InputLayer)                [(None, 10)]                    0           
#> ________________________________________________________________________________
#> dense (Dense)                       (None, 1)                       11          
#> ================================================================================
#> Total params: 11
#> Trainable params: 11
#> Non-trainable params: 0
#> ________________________________________________________________________________
```

Here are a few other suggestions: https://stackoverflow.com/questions/54194724/how-to-use-keras-layers-in-custom-keras-layer. 
However I am seeing an error when trying to train this model with TF 2.3, so not closing this issue yet.

**Comment by dfalbel** [comment]
This works with TF 2.2, I'll file a new issue to track the problem with 2.3 with a smaller reproducible example.
